common/lib/util_sys.c

```diff
@@ -83,113 +75,124 @@ __weak int pal_submit_12v_cycle_slot()
 	return NOT_SUPPORT_12V_CYCLE_SLOT;
 }
 
-#define ME_FW_RECOVERY 0x01
-#define ME_FW_RESTORE 0x02
-
-static void set_ME_FW_mode(uint8_t me_fw_mode)
+/* The byte-2 of ME response for "Get Self-Test Result" command */
+enum GET_SELF_TEST_RESULT_RESPONSE {
+	NO_ERROR = 0x55,
+	FIRMWARE_ENTERED_RECOVERY_MODE = 0x81,
+};
+
+enum RECOVERY_MODE_CAUSE {
+	RECOVERY_MODE_CAUSE_RECOVERY_JUMPER = 0,
+	RECOVERY_MODE_CAUSE_SECURITY_STRAP_OVERRIDE_JUMPER,
+	RECOVERY_MODE_CAUSE_IPMI_COMMAND,
+	RECOVERY_MODE_CAUSE_INVALID_FLASH_CONFIG,
+	RECOVERY_MODE_CAUSE_INTERNAL_ERROR
+};
+
+void set_me_firmware_mode(uint8_t me_fw_mode)
 {
-	ipmi_msg *me_msg;
-	ipmb_error status;
-
-	me_msg = (ipmi_msg *)malloc(sizeof(ipmi_msg));
-	if (me_msg == NULL) {
-		printf("[%s] Failed to allocate memory\n", __func__);
+	if ((me_fw_mode != ME_FW_RECOVERY) && (me_fw_mode != ME_FW_RESTORE)) {
+		printf("Not support the ME frimware mode 0x%x setting", me_fw_mode);
 		return;
 	}
 
-	me_msg->seq_source = 0xFF;
-	me_msg->netfn = NETFN_NM_REQ;
-	me_msg->cmd = 0xDF; // Get Intel ME FW Capabilities
-	me_msg->InF_source = SELF;
-	me_msg->InF_target = ME_IPMB;
-	me_msg->data_len = 4;
-	me_msg->data[0] = 0x57;
-	me_msg->data[1] = 0x01;
-	me_msg->data[2] = 0x00;
-	me_msg->data[3] = me_fw_mode;
-	status = ipmb_read(me_msg, IPMB_inf_index_map[me_msg->InF_target]);
-
-	if (status != IPMB_ERROR_SUCCESS) {
-		printk("set_ME_FW_mode reach ME fail with status: %x\n", status);
+	ipmi_msg *me_msg = (ipmi_msg *)malloc(sizeof(ipmi_msg));
+	uint8_t *data = (uint8_t *)malloc(5 * sizeof(uint8_t));
+	if ((me_msg == NULL) || (data == NULL)) {
+		printf("[%s] Failed to allocate memory\n", __func__);
+		goto cleanup;
 	}
-
-	free(me_msg);
-	return;
-}
-
-void ME_enter_recovery()
-{
-	ipmi_msg *me_msg;
-	ipmb_error status;
-	uint8_t retry = 3;
-
-	for (int i = 0; i < retry; i++) {
-		set_ME_FW_mode(ME_FW_RECOVERY);
-		k_msleep(2000);
-
-		me_msg = (ipmi_msg *)malloc(sizeof(ipmi_msg));
-		if (me_msg == NULL) {
-			printf("ME recovery msg alloc fail\n");
-			k_msleep(10);
+	ipmb_error ret;
+	uint16_t data_len = 0;
+	uint8_t seq_source = 0xFF;
+	int i = 0, retry = 3;
+
+	for (i = 0; i < retry; i++) {
+		seq_source = 0xFF;
+		data_len = 4;
+		data[0] = INTEL_IANA & 0xFF;
+		data[1] = (INTEL_IANA >> 8) & 0xFF;
+		data[2] = (INTEL_IANA >> 16) & 0xFF;
+		data[3] = me_fw_mode;
+		*me_msg = construct_ipmi_message(seq_source, NETFN_NM_REQ,
+						 CMD_OEM_NM_FORCE_ME_RECOVERY, SELF, ME_IPMB,
+						 data_len, data);
+		ret = ipmb_read(me_msg, IPMB_inf_index_map[me_msg->InF_target]);
+		if (ret != IPMB_ERROR_SUCCESS) {
+			printf("Failed to set ME frimware mode to 0x%x, ret: 0x%x\n", me_fw_mode,
```

Comment:
ME firmware

Reply:
We have fixed this and done the forced update.
Please review it. Thanks.